### PR TITLE
dnsmasq: use native 'interface-name' option for local host

### PR DESCRIPTION
**Pull request by EricLuehrsen**
- add_local_host UCI uses dnsmasq instance host file entry
- add_local_host implementation may drop some addresses (i.e. IP6)
- dnsmasq provides this function natively with 'interface-name'
- change add_local_host UCI to act during DHCP assignment
  - use 'interface-name' to assign host to all addresses on interface
  - assign '\<iface\>.\<host\>.\<domain\>' as the true name (PTR)
  - assign '\<host\>.\<domain\>' and '\<host\>' as convinience aliases (not tech. cnames)
- include new UCI add_local_gatename with similar function
  - most ISP don't rDNS (well), so logs/graphs nicer to read
  - default is off for those with dynamic DNS
  - infer WAN as DHCP 'ignore'

Signed-off-by: Eric Luehrsen <ericluehrsen@hotmail.com>

### CODE REVIEW

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 1]
```diff
@@ -380,6 +394,16 @@ dhcp_add() {
 	config_get options "$cfg" options
 	config_get_bool dynamicdhcp "$cfg" dynamicdhcp 1
 
+	if [ $ADD_LOCAL_HOSTNAME -gt 0 ] ; then
+		local dashedname="$( echo $ifname | sed 's/\./-/g' )"
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: You can use `${ifname//./-}` here to avoid the need for spawning a `sed` subshell.

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 2]
```diff
@@ -380,6 +394,16 @@ dhcp_add() {
...
+		local routername="$( uci_get system @system[0] hostname OpenWrt )"
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Hardcoding `OpenWrt` (or `LEDE`) here is not optimal. I suggest to source `/etc/os-release` and to use the `$ID` variable with a fall back to `OpenWrt`.

Example:

    local distribution="$( [ -f /etc/os-release ] && (. /etc/os-release; echo "$ID") || echo openwrt )"
    local routername="$( uci_get system @system[0] hostname "$distribution" )"

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 3]
```diff
@@ -362,7 +363,20 @@ dhcp_add() {
 		DNS_SERVERS="$DNS_SERVERS $dnsserver"
 	}
 
-	append_bool "$cfg" ignore "--no-dhcp-interface=$ifname" && return 0
+	append_bool "$cfg" ignore "--no-dhcp-interface=$ifname" && {
+		if [ $ADD_LOCAL_GATENAME -gt 0 ] ; then
+			local dashedname="$( echo $ifname | sed 's/\./-/g' )"
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: You can use `${ifname//./-}` here to avoid the need for spawning a `sed` subshell.

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 4]
```diff
@@ -362,7 +363,20 @@ dhcp_add() {
...
+			local routername="$( uci_get system @system[0] hostname OpenWrt )"
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Hardcoding `OpenWrt` (or `LEDE`) here is not optimal. I suggest to source `/etc/os-release` and to use the `$ID` variable with a fall back to `OpenWrt`.

Example:

    local distribution="$( [ -f /etc/os-release ] && (. /etc/os-release; echo "$ID") || echo openwrt )"
    local routername="$( uci_get system @system[0] hostname "$distribution" )"

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 5]
```diff
@@ -23,22 +24,22 @@ xappend() {
...
+	# convert into host id
+	export "$var=$(
+		printf "%0x:%0x" \
+		$(((0x$hex >> 16) % 65536)) \
+		$(( 0x$hex        % 65536))
+		)"
 
-        return 0
+	return 0
 }
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Please move those white space changes into a separate commit.

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 6]
```diff
@@ -8,6 +8,7 @@ PROG=/usr/sbin/dnsmasq
 
 ADD_LOCAL_DOMAIN=1
 ADD_LOCAL_HOSTNAME=1
+ADD_LOCAL_GATENAME=0
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: What is the meaning of `GATENAME`? Short hand for gateway hostname? If yes, why is it added for every non-dhcp interface?

Do we even need the distinction? You could make `ADD_LOCAL_HOSTNAME` control adding host names for both DHCP and no-DHCP interfaces.

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 7]
```diff
@@ -380,6 +394,16 @@ dhcp_add() {
...
+	if [ $ADD_LOCAL_HOSTNAME -gt 0 ] ; then
+		local dashedname="$( echo $ifname | sed 's/\./-/g' )"
+		local routername="$( uci_get system @system[0] hostname OpenWrt )"
+		# Put the router host name on this DHCP served interface address(es)
+		# All IP6 addresses discovered by dnsmasq will be labeled (except fe80::)
+		xappend "--interface-name=$dashedname.$routername.$DOMAIN,$ifname"
+		xappend "--interface-name=$routername.$DOMAIN,$ifname"
+		xappend "--interface-name=$routername,$ifname"
+	fi
+
```
Comment (jow-) [inline package/network/services/dnsmasq/files/dnsmasq.init]: This code block looks very similar to the "gatename" handling block above. Maybe you should consider moving it to a procedure?

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 5 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Sorry. Pushed a regex fix of my own mess.

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 7 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Yeah. Until recently my own test of the other was `gateway.out` but now that I have a sensible pattern...

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 2 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: -above-

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 4 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Its what was there. I didnt like it either, but it wasnt material to my goals.

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 1 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: -above-

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 3 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: I get frustrated with so many (non std) shell implementations, that I take the long way first to guarantee function.

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 6 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Yes, better explanation. I took `LOCAL_HOSTNAME` to mean your internal (business) network. If your exterior WAN however had a good name (or a few) in global DNS, then you may not want that sides name changed by this option. That would be for business class service from ISP. `LOCAL_GATENAME` allows you to split it. For most people with residential service on a consumer router, you will just set both true.

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 7 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Until recently my own test of the other was gateway.out but now that I have a sensible pattern... **DONE**

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 4 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Its what was there. I didn't like it either, but it wasn't material to my goals. A fixed name or distro name has name collision potential still. Lets smash the MAC in there. **DONE**

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 3 above)
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: I get frustrated with so many (non std) shell implementations, that I take the long way first to guarantee function. DONE

**Comment by EricLuehrsen** [comment]
@jow- All the requested changes have been fixed or explained. Please advise on next steps.

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 8]
```diff
@@ -604,6 +631,7 @@ dnsmasq_start()
 
 	config_get_bool ADD_LOCAL_DOMAIN "$cfg" add_local_domain 1
 	config_get_bool ADD_LOCAL_HOSTNAME "$cfg" add_local_hostname 1
+	config_get_bool ADD_LOCAL_GATENAME "$cfg" add_local_gatename 0
```
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Yes `add_wan_hostname` would be parallel. For 90%, **add_** is most meaningful, because ISP often dont bother with IP4 `c64-100-57-7.city.state.example.net` and never IP6 due to the wide complexity.

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 9]
```diff
@@ -292,6 +293,25 @@ dhcp_host_add() {
...
 }
 
+dhcp_this_host_add() {
+	# TODO: case-in do/not short-host or FQDN; UCI already intended ...
+	local ifdashname="$1"
+	local do_enable="$2"
+	local routerstub routername
+	
+	ifdashname="${ifdashname//./-}"
+	routerstub="$( cat /sys/class/net/br-lan/address )"
```
Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: oops... left my hard code for a quick test.

Comment (EricLuehrsen) [inline package/network/services/dnsmasq/files/dnsmasq.init]: Oops. That wasn't the best choice either. If no host name (what?), then just MD5 /etc/os-release.

**Comment by dedeckeh** [comment]
@EricLuehrsen Patch fails to apply on latest trunk; can you rebase it ?
Could you make the commit comments of the second patch a bit more self explaining ?

**Comment by EricLuehrsen** [comment]
**FIXED!**

**Comment by EricLuehrsen** [comment]
**Fixed**

**Comment by dedeckeh** [comment]
Pulled into my staging tree; thanks

**Comment by stintel** [comment]
@EricLuehrsen this causes problems when a FQDN is configured in /etc/config/system. The domain name now appears twice in reverse DNS. Can you send a follow-up PR to fix this?

Next to that, there seems to be a bug in dnsmasq. From the manual page:
```
--interface-name=<name>,<interface>[/4|/6]
Return  a  DNS  record  associating  the  name  with  the primary address on the given interface. This flag specifies an A or AAAA record for the given name in the same way as an /etc/hosts line, except that the address is not constant, but taken from the given interface. The interface may be followed by "/4" or "/6" to specify  that  only  IPv4  or  IPv6 addresses of the interface should be used. If the interface is down, not configured or non-existent, an empty record is returned. The matching PTR record is also created, mapping the interface address to the name. More than one name may be associated with an interface address by repeating the flag; in that case the first instance is used for  the  reverse address-to-name mapping.
```
It does not just create an A/AAAA record for the primary address, it creates one for all addresses. And what is worse, it seems to actually resolve to the non-primary address first. This is quite annoying when you use floating IP addresses (e.g. VRRP), because when the floating IP is on the other device, SSH failes due to incorrect entry in the known hosts file.

I know that this is not a common setup, but it would be nice if there was an option to restore the previous behaviour, rather than just forcing this new feature on everybody.

**Comment by EricLuehrsen** [comment]
For issue (1) there are few possible fixes and some more robust for wider array of users. I guess I never though of host name as FQDN; normally you let the domain controller append that for you. (A) if host name has dots, don't add anything to it (B) look for match from end, and squelch the overlap (C) use the UCI which does this also for DHCP and /etc/hosts ... (D) ?

For issue (2) I am not clear on the problem. The old way got the IP4 and list of IP6 addresses, ran a for-do, and labeled **everything** with no filters. dnsmasq does the same thing. It seems transparent.

**Comment by EricLuehrsen** [comment]
I was editing my response when github went down. I have no idea what nonsense I left behind. I guess I will update later.



- Eric


-------- Original message --------

**Comment by dedeckeh** [comment]
@EricLuehrsen As the commit introduces regression issues and since we have a release coming up the best way to go forward is to revert the commit for the moment.
After the release branch off a new PR can be integrated fixing those issues.
